File: shared/src/commonMain/kotlin/com/aptrade/shared/domain/PortfolioPerformance.kt
package com.aptrade.shared.domain

import com.ionspin.kotlin.bignum.decimal.BigDecimal

/**
 * Transcribed from `Sources/APTradeDomain/PortfolioPerformance.swift` (`performanceSeries`).
 *
 * ALL-PRICED GATE (kmp-portfolio-6b2, user-mandated; ADOPTED by macOS in 6b.3): the Swift
 * original once silently dropped any symbol that had no `lastClose` yet at a given union date,
 * so a leading window before a mixed-calendar symbol's first candle (e.g. crypto trading 24/7
 * alongside equities that only have market-hours candles) still emitted a point valuing only
 * the already-priced symbols — producing a valuation cliff the moment the late symbol's first
 * candle joined. This Kotlin implementation instead gates the curve: it emits nothing until
 * every symbol with any history is priced, so the curve starts flat at the first fully-priced
 * date. As of increment 6b.3 the Swift `performanceSeries` adopts this exact gate (reversing
 * the original macOS-first direction), so the two implementations are back in lockstep.
 */

/** One point on a reconstructed portfolio performance curve: the account's value and its
 *  unrealized P&L at a past date, valued against that date's market prices. */
data class PortfolioPerformancePoint(
    val epochSeconds: Long,
    /** Cash plus the market value of current holdings at this date. */
    val value: Money,
    /** Unrealized P&L of current holdings versus average cost at this date. */
    val pnl: Money,
)

/** Reconstructs a value / unrealized-P&L time series by valuing the *current* holdings
 *  against each symbol's historical closes, forward-filling each symbol's last known close
 *  at every date so symbols with different trading calendars (e.g. crypto vs. equities)
 *  stay aligned. Cash is treated as constant. Pure — callers supply the per-symbol
 *  histories; this does no networking. */
fun Portfolio.performanceSeries(histories: Map<String, List<PricePoint>>): List<PortfolioPerformancePoint> {
    if (positions.isEmpty()) return emptyList()
    val code = cash.currencyCode

    val sorted = mutableMapOf<String, List<PricePoint>>()
    val allDates = mutableSetOf<Long>()
    for (position in positions) {
        val points = (histories[position.asset.symbol] ?: emptyList()).sortedBy { it.epochSeconds }
        if (points.isEmpty()) continue
        sorted[position.asset.symbol] = points
        for (point in points) allDates.add(point.epochSeconds)
    }
    if (allDates.isEmpty()) return emptyList()

    val cursor = mutableMapOf<String, Int>()        // forward-fill pointer per symbol
    val lastClose = mutableMapOf<String, BigDecimal>() // last close seen at or before the date
    val result = mutableListOf<PortfolioPerformancePoint>()

    for (date in allDates.sorted()) {
        var holdings = BigDecimal.ZERO
        var pnl = BigDecimal.ZERO
        var allPriced = true
        for (position in positions) {
            val symbol = position.asset.symbol
            val points = sorted[symbol] ?: continue  // no history at all: excluded, doesn't gate
            var i = cursor[symbol] ?: 0
            while (i < points.size && points[i].epochSeconds <= date) {
                lastClose[symbol] = points[i].close.amount
                i += 1
            }
            cursor[symbol] = i
            val close = lastClose[symbol]
            if (close == null) {
                // Gate: this symbol has history but no close yet at this date. Per the
                // recorded divergence above, skip the whole date rather than valuing only
                // the already-priced symbols.
                allPriced = false
                continue
            }
            val quantity = position.quantity
            holdings += close * quantity
            pnl += (close - position.averageCost.amount) * quantity
        }
        if (!allPriced) continue
        result.add(
            PortfolioPerformancePoint(
                epochSeconds = date,
                value = Money(cash.amount + holdings, code),
                pnl = Money(pnl, code),
            ),
        )
    }
    return result
}

File: Sources/APTradeDomain/PortfolioPerformance.swift
import Foundation

/// One point on a reconstructed portfolio performance curve: the account's value and its
/// unrealized P&L at a past date, valued against that date's market prices.
public struct PortfolioPerformancePoint: Equatable, Sendable {
    public let date: Date
    /// Cash plus the market value of current holdings at this date.
    public let value: Money
    /// Unrealized P&L of current holdings versus average cost at this date.
    public let pnl: Money

    public init(date: Date, value: Money, pnl: Money) {
        self.date = date
        self.value = value
        self.pnl = pnl
    }
}

public extension Portfolio {
    /// Reconstructs a value / unrealized-P&L time series by valuing the *current* holdings
    /// against each symbol's historical closes, forward-filling each symbol's last known
    /// close at every date so symbols with different trading calendars (e.g. crypto vs.
    /// equities) stay aligned. Cash is treated as constant. Pure — callers supply the
    /// per-symbol histories; this does no networking.
    ///
    /// All-priced gate (adopted from the Kotlin shared core in increment 6b.3, reversing the
    /// original macOS-first direction): a date is emitted only once *every* symbol with any
    /// history is priced. A symbol that has history but no close yet at a given date sets
    /// `allPriced = false` and the whole date is skipped, so a leading window before a
    /// mixed-calendar symbol's first candle (e.g. crypto trading 24/7 alongside market-hours
    /// equities) no longer emits a point valuing only the already-priced symbols — that
    /// avoided a valuation cliff the moment the late symbol's first candle joins. Symbols
    /// with NO history at all stay excluded from the gate (they must not blank the curve).
    func performanceSeries(histories: [String: [PricePoint]]) -> [PortfolioPerformancePoint] {
        guard !positions.isEmpty else { return [] }
        let code = cash.currencyCode

        var sorted: [String: [PricePoint]] = [:]
        var allDates = Set<Date>()
        for position in positions {
            let points = (histories[position.asset.symbol] ?? []).sorted { $0.date < $1.date }
            guard !points.isEmpty else { continue }
            sorted[position.asset.symbol] = points
            for point in points { allDates.insert(point.date) }
        }
        guard !allDates.isEmpty else { return [] }

        var cursor: [String: Int] = [:]        // forward-fill pointer per symbol
        var lastClose: [String: Decimal] = [:]  // last close seen at or before the date
        var result: [PortfolioPerformancePoint] = []
        result.reserveCapacity(allDates.count)

        for date in allDates.sorted() {
            var holdings = Decimal(0)
            var pnl = Decimal(0)
            var allPriced = true
            for position in positions {
                let symbol = position.asset.symbol
                guard let points = sorted[symbol] else { continue }  // no history at all: excluded, doesn't gate
                var i = cursor[symbol] ?? 0
                while i < points.count, points[i].date <= date {
                    lastClose[symbol] = points[i].close.amount
                    i += 1
                }
                cursor[symbol] = i
                guard let close = lastClose[symbol] else {
                    // Gate: this symbol has history but no close yet at this date. Skip the
                    // whole date rather than valuing only the already-priced symbols.
                    allPriced = false
                    continue
                }
                let quantity = position.quantity.amount
                holdings += close * quantity
                pnl += (close - position.averageCost.amount) * quantity
            }
            guard allPriced else { continue }
            result.append(PortfolioPerformancePoint(
                date: date,
                value: Money(amount: cash.amount + holdings, currencyCode: code),
                pnl: Money(amount: pnl, currencyCode: code)
            ))
        }
        return result
    }
}
